subsys/logging/log_backend_cavs_hda.c

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log_backend.h>
+#include <logging/log_core.h>
+#include <logging/log_msg.h>
+#include <logging/log_output.h>
+#include <logging/log_backend_std.h>
+#include <logging/log_backend_cavs_hda.h>
+#include <drivers/dma.h>
+#include <zephyr/kernel.h>
+
+static uint32_t log_format_current = CONFIG_LOG_BACKEND_CAVS_HDA_OUTPUT_DEFAULT;
+static const struct device *hda_log_dev;
+static uint32_t hda_log_chan;
+
+/*
+ * HDA requires 128 byte aligned data and 128 byte aligned transfers.
+ */
+static __aligned(128) uint8_t hda_log_buf[CONFIG_LOG_BACKEND_CAVS_HDA_SIZE];
+static volatile uint32_t hda_log_buffered;
+static struct k_spinlock hda_log_lock;
+static struct k_timer hda_log_timer;
+static cavs_hda_log_hook_t hook;
+
+/* atomic bit flags for state */
+#define HDA_LOG_DMA_READY 0
+#define HDA_LOG_PANIC_MODE 1
+static atomic_t hda_log_flags;
+
+static uint32_t hda_log_flush(void)
+{
+	uint32_t nearest128 = hda_log_buffered & ~((128) - 1);
+
+	if (nearest128 > 0) {
+		hda_log_buffered = hda_log_buffered - nearest128;
+		dma_reload(hda_log_dev, hda_log_chan, 0, 0, nearest128);
+	}
+
+	return nearest128;
+}
+
+
+static int hda_log_out(uint8_t *data, size_t length, void *ctx)
+{
+	int ret;
+	bool do_log_flush;
+	struct dma_status status;
+
+	k_spinlock_key_t key = k_spin_lock(&hda_log_lock);
+
+	/* Defaults when DMA not yet initialized */
+	uint32_t dma_free = sizeof(hda_log_buf);
+	uint32_t write_pos = 0;
+
+	if (atomic_test_bit(&hda_log_flags, HDA_LOG_DMA_READY)) {
+		ret = dma_get_status(hda_log_dev, hda_log_chan, &status);
+		if (ret != 0) {
+			ret = length;
+			goto out;
+		}
+
+		/* The hardware tells us what space we have available, and where to
+		 * start writing. If the buffer is full we have no space.
+		 */
+		if (status.free <= 0) {
+			ret = length;
+			goto out;
+		}
+
+		dma_free = status.free;
+		write_pos = status.write_position;
+	}
+
+	/* Account for buffered writes since last dma_reload
+	 *
+	 * No underflow should be possible here, status.free is the apparent
+	 * free space in the buffer from the DMA's read/write positions.
+	 * When dma_reload is called status.free may be reduced by
+	 * the nearest 128 divisible value of hda_log_buffered,
+	 * where hda_log_buffered is then subtracted by the same amount.
+	 * After which status.free should only increase in value.
+	 *
+	 * Assert this trueth though, just in case.
+	 */
+	__ASSERT_NO_MSG(dma_free > hda_log_buffered);
+	uint32_t available = dma_free - hda_log_buffered;
+
+	/* If there isn't enough space for the message there's an overflow */
```

Comment:
It occurs to me that this is a situation where you might plausibly want a short write kind of behavior, returning the partial number of bytes consumed.  I have no idea if the log layer is prepared to handle that sort of thing cleanly, but I know it will emit "N messages dropped" kind of diagnostics on the other side (insertion into the async queue), so it's a feature the subsystem might evolve toward in the future.